infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -62,159 +78,160 @@ public class SequencerServerCache implements AutoCloseable {
+            }
+        }
+
+        log.info("evictUpTo: entries={} addresses={}", numConflictKeysEvicted, numTxVersionsToEvict);
+        MicroMeterUtils.measure(numConflictKeysEvicted, "sequencer.cache.evictions");
     }
 
     /**
-     * The memory space used by the entries and also the space used by
-     * priority queue and hashmap to store the pointers
-     * @return the memory space used in bytes:
+     * The cache size as the number of entries.
+     * @return The current cache size.
      */
-    public long byteSize() {
-        log.debug("the cache has {} entries,  the object size used {}, calculated by beepSize {}",
-                size(), CONFLICTTXSTREAM_OBJ_SIZE,
-                cacheEntries.isEmpty() ? 0 : Memory.sizeOf.deepSizeOf(cacheEntries.peek()));
-        return size() * (ENTRY_OVERHEAD + CONFLICTTXSTREAM_OBJ_SIZE);
+    public int size() {
+        return conflictKeyMap.size();
     }
 
-    /*
-     * Put a value in the cache
-     *
-     * @param conflictStream conflict stream
+    /**
+     * Put the provided hashed conflict keys into the cache,
+     * and evict older entries as necessary.
+     * @param conflictKeys The conflict keys to add into the cache.
+     * @param txVersion    The timestamp associated with these conflict keys.
      */
-    public boolean put(ConflictTxStream conflictStream) {
-
-        Long val = conflictKeys.getOrDefault(conflictStream, Address.NON_ADDRESS);
-        if (val > conflictStream.txVersion) {
-            log.error("For key {} the new entry address {} is smaller than the entry " +
-                            "address {} in cache. There is a sequencer regression.",
-                    conflictStream, conflictStream.txVersion, conflictKeys.get(conflictStream));
-            return false;
-        }
+    public void put(@NonNull Set<ConflictTxStream> conflictKeys, long txVersion) {
+        // The provided timestamp should not already be present in the cache.
+        Preconditions.checkState(!versionMap.containsKey(txVersion),
+                "txVersion=%s is already present in SequencerServerCache", txVersion);
 
-        cacheEntries.add(conflictStream);
-        conflictKeys.put(conflictStream, conflictStream.txVersion);
+        // The provided timestamp should be greater than the smallest address in the cache.
+        final long smallestAddress = firstAddress();
+        Preconditions.checkArgument(txVersion > smallestAddress,
+                "txVersion=%s should be larger than the smallest address=%s", txVersion, smallestAddress);
 
-        while (conflictKeys.size() > cacheSize) {
-            invalidateSmallestTxVersion();
-        }
-        return true;
+        // The provided timestamp should be greater than any previous timestamp associated provided conflict
+        // key information. This check is performed separately to avoid partially updating the cache.
+        conflictKeys.parallelStream().forEach(conflictKey -> {
```

Comment:
conflictKeys is a `HashSet` and not thread-safe. Although this just dispatches threads to do reads, I would suggest not adding it unless you can demonstrate performance gain.

Reply:
These checks have been subsumed by other precondition checks after refactoring/reworking.